content/courses/ada-in-practice/chapters/streams_device_driver_api_flexibility.rst

```diff
@@ -0,0 +1,665 @@
+array component assigned in :ada:`Read.Buffer`. The function result is
+:ada:`Buffer'First` |mdash| when no components are assigned, except when that
+would be less than the lowest possible array index value.
+
+Here is a demonstration procedure to be run on the STM32 F4 Discovery Board:
+
+.. code-block:: ada
+
+    with Last_Chance_Handler;  pragma Unreferenced (Last_Chance_Handler);
+    with Serial_IO;
+    with Peripherals_Streaming; use Peripherals_Streaming;
+
+    procedure Demo_Serial_Port_Streaming is
+    begin
+       Serial_IO.Initialize_Hardware (Peripheral);
+       Serial_IO.Configure (COM.Device, Baud_Rate => 115_200);
+       --  This baud rate selection is entirely arbitrary. Note that you may
+       --  have to alter the settings of your host serial port to match this
+       --  baud rate, or just change the above to match whatever the host
+       --  serial port has set already. An application such as TerraTerm
+       --  or RealTerm is helpful.
+
+       loop
+          declare
+             --  await the next msg from the serial port
+             Incoming : constant String := String'Input (COM'Access);
+          begin
+             --  echo the received msg content
+             String'Output (COM'Access, "You sent '" & Incoming & "'");
+          end;
+       end loop;
+    end Demo_Serial_Port_Streaming;
+
+The specific USART on the STM32 F4 Discovery Board must be connected to a
+serial port on the host computer. With that connection in place the embedded
+ARM board and the host computer can communicate over the two serial ports. This
+demonstration iteratively receives a string sent from the host, prepends some
+text, and sends that back, in effect echoing the host sender's text.
+
+The stream attributes :ada:`String'Output` and :ada:`String'Input` write and
+read the bounds as well as the characters. As a consequence, you will need to
+use a program on the host that handles those bounds. A good way to do that is
+to use a host program that also uses streams to send and receive :ada:`String`
+values. Note that the ADL serial port examples include a host application that
+you can build and run for this purpose.
+
+
+Pros
+----
+
+The stream-based approach has all the advantages of the generic-based approach
+without requiring generic instantiations. There is no limit to the number and
+kinds of client types supported, including limited types if they have the
+attributes defined. It is type-safe by default, because the compiler will
+verify that the type used to invoke a stream attribute is the same type as the
+value involved. In addition, the low-level implementation details are hidden
+inside the package body.
+
+Additionally, the approach is independent of other design considerations, such
+as whether callers wait for completion of the invoked I/O operation. The bodies
+of the :ada:`Read` and :ada:`Write` procedures can be implemented to provide
+the expected behavior.
```

Comment:
Let's delete that last sentence above, ie the one reading:
The bodies of the :ada:`Read` and :ada:`Write` procedures can be implemented to provide
the expected behavior.